procinterrupts.c

```diff
@@ -143,6 +143,36 @@ static void guess_arm_irq_hints(char *name, struct irq_info *info)
 	
 	
 }
+
+static int is_arm_irq_aff_cannot_change(int irq)
+{
+	char buf[PATH_MAX] = { 0 };
+	FILE *file = NULL;
+	char *line = NULL;
+	size_t size = 0;
+	int ret = 0;
+
+	snprintf(buf, PATH_MAX - 1, "/proc/irq/%i/smp_affinity", irq);
+	file = fopen(buf, "r+");
+	if (!file)
+		return -1;
+
+	if (getline(&line, &size, file) <= 0) {
+		ret = -1;
+		goto out;
+	}
+
+	line[strlen(line) - 1] = '\0';
+
+	fprintf(file, "%s", line);
+	ret = fflush(file);
+
```

Comment:
And what is this for?  Why are you writing the same affinity back to the file that you just read it from?  I could sort of understand if you tried to write this value and decided that that a failure in the write operation implied an inability to set the affinity, but all smp_affinity files are read/write for root, so what you would really need to do is look and see if a given irqs smp_affinity changes after you write it.

I'm sorry, I have to ask, are you testing this?

Reply:
I'm testing this in arm server, in which each cpu has a irq called mpt3sas0-msix*. For example, cpu1 has a irq called mpt3sas0-msix1 and its number is 13. In /proc/interrupts:
13:          0   17009314          0          0          0          0          0          0 (omit many zeros) ITS-MSI 575143937 Edge      mpt3sas0-msix1
# cat /proc/irq/13/smp_affinity
00000000,00000002
# ll /proc/irq/13/smp_affinity
-rw-r--r-- 1 root root 0 Mar 13 10:04 /proc/irq/13/smp_affinity
it seem like that smp_affinity is writeable, however is_arm_irq_aff_cannot_change(13) return -1 because write smp_affinity  back and fflush fail. when test arch_timer irq, which smp_affinity is ffffffff,ffffffff, the result is the same. When test normal irqs, like ACPI:Ged or ehci_hcd:usb1, it will success and return 0. And i run the test program as root.
If i write it another smp_affinity  and check if irq's smp_affinity changes, when the irq is normal irq, it will success but it will affect irq and system. i think read old smp_affinity -> write new smp_affinity  -> read new smp_affinity -> write old smp_affinity to restore is too redundancy

Reply:
So thats some interesting information.  mpt3sas is an HBA driver, and there should be no reason that the affinity of that devices interrupt is immovable.  The fact that it is suggests a problem in the kernel (though that would surprise me).  What errno value is being returned from fflush?

Additionally, you hit on a real problem with this method of checking for immutable affinity - as I noted, the only way to determine if an irq can't be moved, is to write an affinity mask different from the current one and then rewind (or reopen the file), and read it back comparing the read to the expected value.  If you do that where you've done it, you will change the affinity of the irq (assuming its mutable), which you don't want to do, so this likely isn't the place to make this check.  A better approach would be to do this passively in activate.c.  You could, in activate_mapping, after you've closed the file, reopen it, read the value back, and compare it to info->assigned_obj->mask.  If they differ, then you've found a condition in which an altered affinity for an irq hasn't changed, and you can mark it as such, by logging the error and optionally adding it to the banned list

Reply:
I don't know how hba works. But i see other irqs that have the same question, such as arm-pmu, they are setted IRQF_NOBALANCING and IRQF_PERCPU， maybe this is the reason

Reply:
this looks much better